tests/Transfer/E2E/Sources/Base.php

```diff
@@ -43,8 +43,89 @@ public function testGetSupportedResources(): void
 
     public function testCache(): void
     {
-        $this->source->registerCache($this->createMock(\Utopia\Transfer\Cache::class));
+        $cache = $this->createMock(\Utopia\Transfer\Cache::class);
+        $this->source->registerCache($cache);
 
         $this->assertNotNull($this->source->cache);
     }
+
+    /**
+     * Call
+     *
+     * Make an API call
+     *
+     * @throws \Exception
+     */
+    protected function call(string $method, string $path = '', array $headers = [], array $params = []): array|string
+    {
+        $ch = curl_init((str_contains($path, 'http') ? $path.(($method == 'GET' && ! empty($params)) ? '?'.http_build_query($params) : '') : $path.(($method == 'GET' && ! empty($params)) ? '?'.http_build_query($params) : '')));
+        $responseHeaders = [];
+        $responseStatus = -1;
+        $responseType = '';
+        $responseBody = '';
+
+        switch ($headers['Content-Type']) {
+            case 'application/json':
+                $query = json_encode($params);
+                break;
+
+            default:
+                $query = http_build_query($params);
+                break;
+        }
+
+        foreach ($headers as $i => $header) {
+            $headers[] = $i.':'.$header;
+            unset($headers[$i]);
+        }
+
+        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+        curl_setopt($ch, CURLOPT_USERAGENT, php_uname('s').'-'.php_uname('r').':php-'.phpversion());
+        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
+        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
+        curl_setopt($ch, CURLOPT_HEADERFUNCTION, function ($curl, $header) use (&$responseHeaders) {
+            $len = strlen($header);
+            $header = explode(':', strtolower($header), 2);
+
+            if (count($header) < 2) { // ignore invalid headers
+                return $len;
+            }
+
+            $responseHeaders[strtolower(trim($header[0]))] = trim($header[1]);
+
+            return $len;
+        });
+
+        if ($method != 'GET') {
+            curl_setopt($ch, CURLOPT_POSTFIELDS, $query);
+        }
+
+        $responseBody = curl_exec($ch);
+
+        $responseType = array_key_exists('Content-Type', $responseHeaders) ?? $responseHeaders['content-type'] ?? '';
```

Comment:
We don't need to use `array_key_exists` here, using the null coalescing `?? ''` will work  even if the key doesn't exist in PHP